gobblin-core/src/main/java/gobblin/writer/PartitionedDataWriter.java

```diff
@@ -228,4 +244,82 @@ private boolean isDataWriterForPartitionSafe(DataWriter dataWriter) {
     return dataWriter instanceof  SpeculativeAttemptAwareConstruct
         && ((SpeculativeAttemptAwareConstruct) dataWriter).isSpeculativeAttemptSafe();
   }
+
+  @Override
+  public boolean isWatermarkCapable() {
+    return this.isWatermarkCapable();
+  }
+
+  @Override
+  public void writeEnvelope(RecordEnvelope<D> recordEnvelope)
+      throws IOException {
+    try {
+      DataWriter<D> writer = getDataWriterForRecord(recordEnvelope.getRecord());
+      // Unsafe cast, presumably we've checked earlier through isWatermarkCapable()
+      // that we are wrapping watermark aware wrappers
+      ((WatermarkAwareWriter) writer).write(recordEnvelope);
+    } catch (ExecutionException ee) {
+      throw new IOException(ee);
+    }
+  }
+
+  @Override
+  public Optional<ComparableWatermark> getCommittableWatermark() {
```

Comment:
Agree that it would be up to the writer to handle such scenarios, but we do need to flesh out the expected behavior. Also, I don't expect there to be many WatermarkAwareWriters. Most writers will just be AsyncDataWriters / SyncDataWriters and just be responsible for writing records and flushing pending writes, so they won't care about this stuff. 
There will possibly be a single WatermarkAwareWriter that wraps them. My current plan to have AsyncWriterManager be that class.